### ISSUE TITLE
Rclone doesn't recongize one drive business

### ISSUE THREAD

**Issue by maspodil** [rclone/rclone#4068]
#### What is the problem you are having with rclone?

Rclone doesn't recognize one drive business

#### What is your rclone version (output from `rclone version`)

rclone-v1.51.0

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

Windows 7, ultimate 64 bit
####  Which cloud storage system are you using? (eg Google Drive)

one drive
#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

I was trying to select one drive as a storage but it keeps showing the below 2 sharepoints
#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

output
Use auto config?
 * Say Y if not sure
 * Say N if you are working on a remote or headless machine
y) Yes (default)
n) No
y/n> y
If your browser doesn't open automatically go to the following link: http://127.
0.0.1:53682/auth?state=QDeD3iUA8Qhl4ZvMO6GBbA
Log in and authorize rclone for access
Waiting for code...
Got code
Choose a number from below, or type in an existing value
 1 / OneDrive Personal or Business
   \ "onedrive"
 2 / Root Sharepoint site
   \ "sharepoint"
 3 / Type in driveID
   \ "driveid"
 4 / Type in SiteID
   \ "siteid"
 5 / Search a Sharepoint site
   \ "search"
Your choice> 1
Found 2 drives, please select the one you want to use:
0: Organization Logos (documentLibrary) id=b!BQiVy3rHcUSduGcNNAd-abs3NtJekRNDt6L
-76P7-O3Jgwq3_sFEQa0g0yiEF9LI
1: User Photos (documentLibrary) id=b!BQiVy3rHcUSduGcNNAd-abs3NtJekRNDt6L-76P7-O
0c9Uov7yp8R6ISqBmmgKjk

**Comment by ncw** [comment]
This is probably because your onedrive administrator hasn't allowed access to files.

**Comment by maspodil** [comment]
It works only if I create my own client id and secret key but if I use the default i.e., rclone's client and secret it will not show the one drive storage instead it will display those 2 storages.
So can you help me out?

**Comment by Dibbyo456** [comment]
> instead it will display those 2 storages.

Yeah, I have suffered the same issue even with my own client id sometimes, specially on newly created account.

If it is a newly created account then create/upload a single file in onedrive via browser, after that try rclone, 90% chance it will work. Whenever I face this issue, doing that solved most of the times.

**Comment by maspodil** [comment]
Since I am not the only one who suffers this bug/problem it's definitely not from my side and I can easily use gdrive without creating any client id and secret,so can anyone fix this?

**Comment by ncw** [comment]
Any ideas? Is this something to do with rclone's default client_id?

**Comment by Dibbyo456** [comment]
> Is this something to do with rclone's default client_id?

No, even with own custom client i get this issue sometimes. I am looking into it.

**Comment by Dibbyo456** [comment]
I finally debugged it.

When we config rclone and choose `OneDrive Personal or Business`, rclone send request to `/me/drives` endpoint. However the `/me/drives` endpoint returns the following response:

```
{
    "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#drives",
    "value": [
        {
            "createdDateTime": "2019-07-27T09:07:55Z",
            "description": "Use this Picture Library to store logos for Organizations.",
            "id": "{redacted}",
            "lastModifiedDateTime": "2019-07-27T09:07:55Z",
            "name": "Organization Logos",
            "webUrl": "{redacted}",
            "driveType": "documentLibrary",
            "createdBy": {
                "user": {
                    "displayName": "System Account"
                }
            },
            "quota": {
                "deleted": 0,
                "remaining": 0,
                "total": 0,
                "used": 0
            }
        },
        {
            "createdDateTime": "2019-07-27T09:07:55Z",
            "description": "Profile Pictures for all users will be stored here.",
            "id": "{redacted}",
            "lastModifiedDateTime": "2020-03-17T21:13:25Z",
            "name": "User Photos",
            "webUrl": "{redacted}",
            "driveType": "documentLibrary",
            "createdBy": {
                "user": {
                    "displayName": "System Account"
                }
            },
            "quota": {
                "deleted": 0,
                "remaining": 0,
                "total": 0,
                "used": 0
            }
        }
    ]
}
```
As you can see the actual drive information (documents root) isn't there.

**How to Fix**:

To fix the issue we need to call `/me/drive` instead of `/me/drives` (Note: there is no `s`)

Here is the response if we call `/me/drive` endpoint:

```
{
    "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#drives/$entity",
    "createdDateTime": "2020-03-07T20:08:42Z",
    "description": "",
    "id": "{redacted}",
    "lastModifiedDateTime": "2020-03-20T07:43:45Z",
    "name": "OneDrive",
    "webUrl": "https://{redacted}-my.sharepoint.com/personal/{redacted}/Documents",
    "driveType": "business",
    "createdBy": {
        "user": {
            "displayName": "System Account"
        }
    },
    "lastModifiedBy": {
        "user": {
            "email": "{redacted}",
            "id": "{redacted}",
            "displayName": "{redacted}"
        }
    },
    "owner": {
        "user": {
            "email": "{redacted}",
            "id": "{redacted}",
            "displayName": "{redacted}"
        }
    },
    "quota": {
        "deleted": 0,
        "remaining": 5497557365266,
        "state": "normal",
        "total": 5497558138880,
        "used": 0
    }
}
```
In short: all we need is remove the `s` from [Line 129](https://github.com/rclone/rclone/blob/master/backend/onedrive/onedrive.go#L129).

**Comment by Cnly** [comment]
@Dibbyo456 Thanks for your time and effort on this! It seems this fix works because `/me/drive` [automatically provisions](https://docs.microsoft.com/en-us/onedrive/developer/rest-api/api/drive_get?view=odsp-graph-online#get-current-users-onedrive) the user's OneDrive:

> If a user's OneDrive is not provisioned but the user has a license to use OneDrive, this request will automatically provision the user's drive, when using delegated authentication.

Would you like to open a PR for this?

**Comment by Dibbyo456** [comment]
@Cnly sorry for removing the previous comment I accidentally posted some info which I needed to remove.

> Would you like to open a PR for this?

Well I can remove `s` but I don't think that is sufficient, because it may need some function to catch the `id`, I don't know golang (although I am learning).

------

**Original comment:** I finally debugged it.

When we config rclone and choose `OneDrive Personal or Business`, rclone send request to `/me/drives` endpoint. However the `/me/drives` endpoint returns the following response:

```
{
    "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#drives",
    "value": [
        {
            "createdDateTime": "2019-07-27T09:07:55Z",
            "description": "Use this Picture Library to store logos for Organizations.",
            "id": "{redacted}",
            "lastModifiedDateTime": "2019-07-27T09:07:55Z",
            "name": "Organization Logos",
            "webUrl": "{redacted}",
            "driveType": "documentLibrary",
            "createdBy": {
                "user": {
                    "displayName": "System Account"
                }
            },
            "quota": {
                "deleted": 0,
                "remaining": 0,
                "total": 0,
                "used": 0
            }
        },
        {
            "createdDateTime": "2019-07-27T09:07:55Z",
            "description": "Profile Pictures for all users will be stored here.",
            "id": "{redacted}",
            "lastModifiedDateTime": "2020-03-17T21:13:25Z",
            "name": "User Photos",
            "webUrl": "{redacted}",
            "driveType": "documentLibrary",
            "createdBy": {
                "user": {
                    "displayName": "System Account"
                }
            },
            "quota": {
                "deleted": 0,
                "remaining": 0,
                "total": 0,
                "used": 0
            }
        }
    ]
}
```
As you can see the actual drive information (documents root) isn't there.

**How to Fix**:

To fix the issue we need to call `/me/drive` instead of `/me/drives` (Note: there is no `s`)

Here is the response if we call `/me/drive` endpoint:

```
{
    "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#drives/$entity",
    "createdDateTime": "2020-03-07T20:08:42Z",
    "description": "",
    "id": "{redacted}",
    "lastModifiedDateTime": "2020-03-20T07:43:45Z",
    "name": "OneDrive",
    "webUrl": "https://{redacted}-my.sharepoint.com/personal/{redacted}/Documents",
    "driveType": "business",
    "createdBy": {
        "user": {
            "displayName": "System Account"
        }
    },
    "lastModifiedBy": {
        "user": {
            "email": "{redacted}",
            "id": "{redacted}",
            "displayName": "{redacted}"
        }
    },
    "owner": {
        "user": {
            "email": "{redacted}",
            "id": "{redacted}",
            "displayName": "{redacted}"
        }
    },
    "quota": {
        "deleted": 0,
        "remaining": 5497557365266,
        "state": "normal",
        "total": 5497558138880,
        "used": 0
    }
}
```
In short: all we need is remove the `s` from [Line 129](https://github.com/rclone/rclone/blob/master/backend/onedrive/onedrive.go#L129).

**Comment by ncw** [comment]
I'm not sure what is going on here...

Here are the docs for the [/me/drives call](https://docs.microsoft.com/en-us/onedrive/developer/rest-api/api/drive_list?view=odsp-graph-online#list-the-current-users-drives) and here are the docs for the [/me/drive call](https://docs.microsoft.com/en-us/onedrive/developer/rest-api/api/drive_get?view=odsp-graph-online#http-request) which returns a single drive for the current user.

I would have thought the data drive would show up in the `drives` list...

If they are going to return different things, then we should probably call `/drive` **and** `/drives` and combine the two (deduplicating on driveID).

**Comment by Dibbyo456** [comment]
Yes, calling both will be the perfect solution.

**Comment by ncw** [comment]
@Dibbyo456 do I understand from your forum post that this is no longer an issue?

**Comment by Dibbyo456** [comment]
No, this is an different issue. Graph API sometimes unable to return onedrive root document folder, because of the onedrive hasn't provisioned yet or the api sending cached response.

> then we should probably call `/drive` and `/drives` and combine the two (deduplicating on driveID).

Doing this probably would be the best idea.

**Comment by ncw** [comment]
I had a go at this - can you have a try?

If you run `rclone config -vv --dump bodies` it will show you the http transactions - please send that if it doesn't work!

https://beta.rclone.org/branch/v1.51.0-126-g7bf37666-fix-4068-onedrive-drives-beta/ (uploaded in 15-30 mins)

**Comment by Dibbyo456** [comment]
Fixed. 👍

![](https://i.imgur.com/pjm8aoV.png)

**Comment by ncw** [comment]
Thanks for testing :-)

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.52
